lib/rust/frp/src/io/timer/delayed_interval.rs

```diff
@@ -0,0 +1,86 @@
+///     - Subsequent `on_trigger` events will be emitted every `interval_ms` time.
+/// Once `stop` receives an event, no more `on_trigger` events will be emitted until the timer
+/// is activated again.
+///
+/// in `start`:       -5,2--------------3,1-----4,3-----------------
+/// in `stop`:        -------------x-----------------------x---x----
+/// out `on_trigger`: ------x-x-x-x--------xxxxx----x--x--x---------
+///
+/// The timer is based on `setTimeout` and `setInterval` browser APIs. That means there is no
+/// guarantee about when exactly the events will be emitted. In practice, the initial delay might be
+/// slightly longer and there might be some jitter in periodic triggers. This effect will be more
+/// pronounced when the browser event loop is very busy.
+#[derive(Clone, CloneRef, Debug)]
+pub struct DelayedInterval {
+    delay_timer:    Timeout,
+    interval_timer: Interval,
+    /// Timer activation and configuration input. Any time an event is sent to this input, the
+    /// timer will be started with provided delay and interval values. If the timer was already
+    /// started, it will be restarted with new values.
+    pub start:      frp::Any<DelayedIntervalConfig>,
+    /// Timer deactivation input. Any time an event is sent to this input, the timer will be
+    /// stopped and mo more `on_trigger` will be emitted until it is started again.
+    pub stop:       frp::Any,
+    /// Emitted after initial delay and then every interval, as long as the timer is active.
+    pub on_trigger: frp::Stream<()>,
+}
+
+impl DelayedInterval {
+    /// Constructor. Timer is initially not active.
+    pub fn new(network: &frp::Network) -> Self {
+        let delay_timer = Timeout::new(network);
+        let interval_timer = Interval::new(network);
+
+        frp::extend! { network
+            start <- any_mut::<DelayedIntervalConfig>();
+            stop <- any_mut();
+
+            delay_timer.start    <+ start.map(|c| c.delay_ms);
+            interval_timer.start <+ start.map(|c| c.interval_ms).sample(&delay_timer.on_expired);
+
+            delay_timer.cancel  <+ stop;
+            interval_timer.stop <+ any_(stop, start);
+
+            on_trigger <- any(delay_timer.on_expired, interval_timer.on_interval);
+        }
+        Self { delay_timer, interval_timer, start, stop, on_trigger }
+    }
+}
+
+/// Configuration structure for `DelayedInterval`.
+/// Specifies the time duration of each phase in integer milliseconds.
+#[derive(Debug, Clone, Copy, Default)]
+pub struct DelayedIntervalConfig {
+    /// Initial delay between timer activation and first `on_trigger` event being emitted.
+    pub delay_ms:    i32,
+    /// Time between subsequent `on_trigger` events.
+    pub interval_ms: i32,
+}
+
+impl DelayedIntervalConfig {
+    /// Constructor.
+    pub fn new(delay_ms: i32, interval_ms: i32) -> Self {
+        Self { delay_ms, interval_ms }
+    }
```

Comment:
Weare almost there. This is kind of anti-pattern with FRP. Instead, these config should be set up by FRP inputs `set_delay` and `set_interval`. In order to do it, use the `define_endpoints_2`, store `Frp` in the `DelayedInterval` and `Deref` to it.

Reply:
The `define_endpoints_2` macro is part of `ensogl_core`, which depends on the `frp` crate. I can't use it here.

I am also not thrilled about this `DelayedIntervalConfig` struct, but I don't really see a great alternative. The main reason it exists it is that you can't change interval values while the timer is running. I wanted an API that makes it very clear that changing the timeout values also restarts the whole timer applying both timeout values at once, and there is no way around that. If I split this into separate `set_delay` and `set_interval` streams, the semantics around timer restarts would be potentially surprising.

Avoiding this whole problem was my motivation for using simple boolean based API before. Unfortunately that led to interval values being essentially static, so I changed it to a stream accepting both values at once. Please let me know if you have better ideas how to handle that case.